exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/ScanBatch.java

```diff
@@ -252,14 +235,28 @@ public IterOutcome next() {
     }
   }
 
+  private void releaseAssets() {
+    container.zeroVectors();
+  }
+
+  private void clearFieldVectorMap() {
+    for (final ValueVector v : mutator.fieldVectorMap().values()) {
+      v.clear();
+    }
+  }
+
+  private void advanceNextReader() throws ExecutionSetupException {
+    currentReader = readers.next();
+    implicitValues = implicitColumns.hasNext() ? implicitColumns.next() : null;
+    currentReader.setup(oContext, mutator);
```

Comment:
Nice catch. You are right we should not advance to next reader if we need pass a batch to downstream, since `reader.setup()` for new reader could modify both mutator and container.  This seems to be true for original ScanBatch code as well. 

In the revised batch,  I remove the code to get next reader in constructor, and put the code to get next reader in the beginning of loop when needed. In case of return a `OK_NEW_SCHEMA`, we will only get next reader in the next `next()` method call after `OK_NEW_SCHEMA` is returned.